### PR TITLE
M3-2442 Add: Events landing page

### PR DESCRIPTION
**Pull request by Jskobos**
## Description

Go to /events to test, or use the View All Events link in the events dropdown (the bell menu)

## Note to Reviewers

- This is a paginated list without any links. Plan is to make it use infinite scroll (so that we can be sure users will load events in reverse chronological order, which in turn allows us to determine if an event's entity link is valid).
- Most of the changed files are test-related; functional changes are mostly in looking at UserEventList, EventsLanding, and EventRow.
- Also extracted some of the events-related utilities (getting link targets) to a separate file, and added a new (curried!) function for retrieving an entity from the store.

### CODE REVIEW

File: src/utilities/getEntityByIDFromStore.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Usage:
+ *
+ * getEntityByIDFromStore('linode', 123456) => Linode | undefined
+ *
+ * The function is curried, so you can make entity-specific functions as needed:
+ *
+ * const getLinode = getEntityByIDFromStore('linode');
+ * getLinode(123456) => Linode | undefined
+ */
+const _getEntityByIDFromStore = (
```
Comment (johnwcallahan) [inline src/utilities/getEntityByIDFromStore.ts]: Love the currying here.

**Comment by WilkinsKa1** [comment]
This is weird, I can only replicate on my account, events don't display in the UI, however I do see them all coming back in the console as unknown events. No errors otherwise:
Screen Shot 2019-03-20 at 2 00 12 PM

**Comment by Jskobos** [comment]
Marking this as do not merge, as if any new API events are added soon (which they might be) the paginated implementation won't work.

File: src/features/Events/EventsLanding.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  const [events, setEvents] = React.useState<Linode.Event[]>([]);
+  const [loading, setLoading] = React.useState<boolean>(false);
+  const [error, setError] = React.useState<string | undefined>(undefined);
+  const [currentPage, setCurrentPage] = React.useState<number>(1);
+  const [isRequesting, setRequesting] = React.useState<boolean>(false);
+
+  const getNext = () => {
+    setRequesting(true);
+    setCurrentPage(currentPage + 1);
+    getEvents({ page: currentPage, pageSize: 50 }).then(
```
Comment (martinmckenna) [inline src/features/Events/EventsLanding.tsx]: is `useState` synchronous? I couldn't find anything in the docs, but I do see that you can do callbacks with it

https://reactjs.org/docs/hooks-reference.html#functional-updates

```js
setCurrentPage(currentPage => {
  getEvents().then(.....)
  return currentPage + 1
})
```

Not sure if this is the correct way of doing things - if `useState` is not async, then my point is moot

**Comment by martinmckenna** [comment]
We should also get rid of that useless eventTypes.ts file

**Comment by martinmckenna** [comment]
Screen Shot 2019-03-27 at 9 47 07 AM

Getting this on initial load

**Comment by alioso** [comment]
@Jskobos I did quite a bit of work in here:
- refactored the top menu dropdown in order to make the link always visible
- fixed the infinite loading so that it won't keep loading if end of data
- styling a few things for the table as well

File: src/features/Events/EventsLanding.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    if (response.pages === currentPage) {
+      setLoadMoreEvents(false);
+    }
+  };
+
+  React.useEffect(() => {
+    setLoading(true);
+    setRequesting(true);
+    setError(undefined);
+    getEvents().then(handleEventsRequestSuccess);
```
Comment (johnwcallahan) [inline src/features/Events/EventsLanding.tsx]: I think we'll need to `setLoading(false)` and set errors in a `catch` block here. Currently, if `/events` fails, we're stuck with a loading spinner.

File: src/features/Events/EventsLanding.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            {renderTableBody(isLoading, isRequesting, events, error)}
+          </TableBody>
+        </Table>
+      </Paper>
+      {loadMoreEvents ? (
+        <Waypoint onEnter={getNext}>
+          <div style={{ minHeight: '150px' }} />
+        </Waypoint>
+      ) : (
+        !loading && (
```
Comment (johnwcallahan) [inline src/features/Events/EventsLanding.tsx]: I wonder if there should be a secondary error state to handle errors on the additional fetching of events. Right now we're stuck with a spinner, but ideally we'd show an error message while keeping the events that have already been loaded intact.

File: src/features/Events/EventRow.test.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import 'jest-dom/extend-expect';
```
Comment (martinmckenna) [inline src/features/Events/EventRow.test.tsx]: is this still necessary if it's being included globally?

File: src/features/Events/EventsLanding.tsx (hunk 4 above)
Comment (martinmckenna) [reply src/features/Events/EventsLanding.tsx]: I agree. Possibly just leverage the `<TableRowError />` and still allow the user to scroll down to trigger the lazy load again

**Comment by alioso** [comment]
@Jskobos i addressed (i think) @johnwcallahan catch comments but am running in sorting/duplication issues for the event list

**Comment by martinmckenna** [comment]
As discussed in Slack

events1

Here you can see that I've loaded the page, scrolled down, and the page requests page 1 of events, even though page one has already been requested

events2

Here I was using the trackpad to quickly scroll and it appears here both pages 3 and 4 are being requested at once (probably because the waypoint was entered twice.

**Pull request merged**